File: common/src/main/java/io/xol/chunkstories/voxel/material/VoxelMaterialsStore.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.voxel.material;

import java.io.BufferedReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.xol.chunkstories.api.content.Asset;
import io.xol.chunkstories.api.content.Content;
import io.xol.chunkstories.api.voxel.materials.VoxelMaterial;
import io.xol.chunkstories.content.GameContentStore;
import io.xol.chunkstories.voxel.VoxelsStore;

public class VoxelMaterialsStore implements Content.Voxels.VoxelMaterials {
	private final GameContentStore store;
	@SuppressWarnings("unused")
	private final VoxelsStore voxels;

	private static final Logger logger = LoggerFactory.getLogger("content.materials");

	public Logger logger() {
		return logger;
	}

	public VoxelMaterialsStore(VoxelsStore voxels) {
		this.voxels = voxels;
		this.store = voxels.parent();
	}

	Map<String, VoxelMaterial> materials = new HashMap<String, VoxelMaterial>();

	public void reload() {
		materials.clear();

		Iterator<Asset> i = store.modsManager().getAllAssetsByExtension("materials");
		while (i.hasNext()) {
			Asset f = i.next();
			readDefinitions(f);
		}
	}

	private void readDefinitions(Asset f) {
		try {
			BufferedReader reader = new BufferedReader(f.reader());
			String line = "";

			VoxelMaterialImplementation material = null;
			while ((line = reader.readLine()) != null) {
				line = line.replace("\t", "");
				if (line.startsWith("#")) {
					// It's a comment, ignore.
				}
				// We shouldn't come accross end tags by ourselves, this is dealt with in the
				// constructors
				else if (line.startsWith("end")) {
					// if (material == null)
					{
						logger().warn("Syntax error in file : " + f + " : ");
						continue;
					}
				} else if (line.startsWith("material")) {
					if (line.contains(" ")) {
						String[] split = line.split(" ");
						String materialName = split[1];

						material = new VoxelMaterialImplementation(materialName, reader);

						// Eventually add the material
						materials.put(material.getName(), material);
					}
				}
			}
			reader.close();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public VoxelMaterial getVoxelMaterial(String name) {
		VoxelMaterial material = materials.get(name);
		if (material != null)
			return material;

		return getVoxelMaterial("undefined");
	}

	@Override
	public Iterator<VoxelMaterial> all() {
		return materials.values().iterator();
	}

	@Override
	public Content parent() {
		return store;
	}
}

File: common/src/main/java/io/xol/chunkstories/voxel/material/VoxelMaterialsStore.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.voxel.material

import DefinitionsLexer
import DefinitionsParser
import java.util.HashMap
import org.slf4j.Logger
import org.slf4j.LoggerFactory

import io.xol.chunkstories.api.content.Asset
import io.xol.chunkstories.api.content.Content
import io.xol.chunkstories.api.voxel.materials.VoxelMaterial
import io.xol.chunkstories.content.GameContentStore
import io.xol.chunkstories.util.format.toMap
import io.xol.chunkstories.voxel.VoxelsStore
import org.antlr.v4.runtime.ANTLRInputStream
import org.antlr.v4.runtime.CommonTokenStream

class VoxelMaterialsStore(private val voxels: VoxelsStore) : Content.Voxels.VoxelMaterials {
    private val store: GameContentStore

    internal var materials: MutableMap<String, VoxelMaterial> = HashMap()

    override val defaultMaterial: VoxelMaterial
        get() = null!!

    override fun logger(): Logger {
        return logger
    }

    init {
        this.store = voxels.parent()
    }

    fun reload() {
        materials.clear()

        fun readDefinitions(a: Asset) {
            logger().debug("Reading voxel material definitions in : $a")

            val text = a.reader().use { it.readText() }
            val parser = DefinitionsParser(CommonTokenStream(DefinitionsLexer(ANTLRInputStream(text))))

            for(definition in parser.voxelMaterialDefinitions().voxelMaterialDefinition()) {
                val name = definition.Name().text
                val properties = definition.properties().toMap()

                val voxelMaterial = VoxelMaterial(this, name, properties)
                materials.put(name, voxelMaterial)

                logger.debug("Loaded voxel material $voxelMaterial")
            }
        }

        val i = store.modsManager().getAllAssetsByExtension("materials")
        while (i.hasNext()) {
            val f = i.next()
            readDefinitions(f)
        }
    }

    override fun getVoxelMaterial(name: String): VoxelMaterial {
        val material = materials[name]
        return material ?: getVoxelMaterial("undefined")

    }

    override fun all(): Iterator<VoxelMaterial> {
        return materials.values.iterator()
    }

    override fun parent(): Content {
        return store
    }

    companion object {

        private val logger = LoggerFactory.getLogger("content.materials")
    }
}
